network/src/peers/peer/outbound_handler.rs

```diff
@@ -0,0 +1,162 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::time::Instant;
+
+use tokio::sync::{mpsc, oneshot};
+
+use crate::{stats, NetworkError, Payload, Peer};
+
+use super::network::PeerNetwork;
+
+pub type PeerJudge = dyn Fn(&Peer) -> bool + Send + Sync + 'static;
+
+pub(super) enum PeerAction {
+    Disconnect,
+    Send(Payload),
+    Get(oneshot::Sender<Peer>),
+    QualityJudgement,
+    CancelSync,
+    GotSyncBlock(oneshot::Sender<bool>),
+    ExpectingSyncBlocks(u32),
+    SoftFail,
+}
+
+#[derive(Clone, Debug)]
+pub struct PeerHandle {
+    pub(super) sender: mpsc::Sender<PeerAction>,
+}
+
+impl PeerHandle {
+    pub async fn load(&self) -> Option<Peer> {
+        metrics::increment_gauge!(stats::QUEUES_OUTBOUND, 1.0);
```

Comment:
the handling of this stat doesn't seem correct; it should only be incremented when a message is queued to be sent to a peer (and decremented when the message has been pushed to the stream)

Reply:
if we are interested in peer-handling queues to also be measured, they need to have their own new metric

Reply:
messages and other actions are in the same queue now, it makes sense to me to keep it in one metric for now, since anything in that queue will have perf implications and possible IO. open to adding more metrics. do you know if `metrics` crate supports tagged metrics? would be super useful here.

Reply:
prometheus can utilize labels: https://prometheus.io/docs/practices/naming/; heads up: the structure of the names is important, otherwise tools like grafana don't recognize them properly